clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -99,6 +99,42 @@ static void initializeAlloca(CodeGenFunction &CGF, AllocaInst *AI, Value *Size,
   I->addAnnotationMetadata("auto-init");
 }
 
+static Value *handleHlslClip(const CallExpr *E, CodeGenFunction *CGF) {
+  Value *Op0 = CGF->EmitScalarExpr(E->getArg(0));
+
+  Constant *FZeroConst = ConstantFP::getZero(CGF->FloatTy);
+  Value *CMP;
+
+  if (const auto *VecTy = E->getArg(0)->getType()->getAs<clang::VectorType>()) {
+    FZeroConst = ConstantVector::getSplat(
+        ElementCount::getFixed(VecTy->getNumElements()), FZeroConst);
+    auto *FCompInst = CGF->Builder.CreateFCmpOLT(Op0, FZeroConst);
+    CMP = CGF->Builder.CreateIntrinsic(
+        CGF->Builder.getInt1Ty(), CGF->CGM.getHLSLRuntime().getAnyIntrinsic(),
+        {FCompInst}, nullptr);
+  } else
+    CMP = CGF->Builder.CreateFCmpOLT(Op0, FZeroConst);
+
+  if (CGF->CGM.getTarget().getTriple().isDXIL())
+    return CGF->Builder.CreateIntrinsic(CGF->VoidTy, llvm::Intrinsic::dx_clip,
```

Comment:
This is a bit different than how we have done intrinsics that invoke other intrinsics. For example see `expandNormalizeIntrinsic` in `llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp`  For example `dx_normaize` exapnds to:
```
fdiv
dx_dot
dx_rsqrt
```
While in this case we associate the instruction expansion with the  clang builtin. This isn't wrong, its just different as past implementers have tried to keep target specific behaviors in their respective targets.

Reply:
I will be happy to move the expansion into `llvm/lib/Target/DirectX/DXILIntrinsicExpansion.cpp`, in order to keep code consistency. However, I have a question, do we have a similar "expassion" pass for spirv? Asking because the `any` intrinsic is used both by SPIRV and DXIL backend in order to handle the vector scenario.

Reply:
For SPIRV I beleive you could just invoke `selectAny` in `llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp`. Also not saying this has to change. But I want to be conservative with adding new patterns unless there is a good reason to do so.

Reply:
Here is an example of the current generated IR:
```
define void @test_vector(<4 x float> noundef %Buf) {
entry:
  %Buf.addr = alloca <4 x float>, align 16
  store <4 x float> %Buf, ptr %Buf.addr, align 16
  %1 = load <4 x float>, ptr %Buf.addr, align 16
  %2 = fcmp olt <4 x float> %1, zeroinitializer
  %3 = call i1 @llvm.spv.any.v4i1(<4 x i1> %2)
  br i1 %3, label %lt0, label %end

lt0:                                              ; preds = %entry
  call void @llvm.spv.clip()
  br label %end

end:                                              ; preds = %lt0, %entry
  ret void
}
```

The challenge, I see, with invoke `selectAny` when handling `selectClip` is that we need to call `any` before calling the conditional branch. In order to call `selectAny` in `llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp` that would also require creating the branch call when selecting the spirv instruction. That doesn't seem to be the proper place to generate the branching code. Please, let me know if I am missing something here.

Reply:
Right now, you have  `@llvm.spv.clip()` maps to `OpDemoteToHelperInvocation`  or `OpKill` depending on some extension availability. What I'm suggesting is twofold. First make your intrinsic map to the same number of arguments as your clang alias. Second that will mean that means register 1 to `selectClip` is what you will be used by `selectAny` Then do your branch invocations only in the spirv backend. In summary the instruction selection will be everything you need for this intrinsic not just OpDemoteToHelperInvocation`  or `OpKill`.

https://hlsl.godbolt.org/z/ose5EGWEx
```
%21 = OpFOrdLessThan %v4bool %20 %11
 %22 = OpAny %bool %21
             OpSelectionMerge %23 None
             OpBranchConditional %22 %24 %23
%24 = OpLabel
            OpKill | OpDemoteToHelperInvocation
 %23 = OpLabel
```

Reply:
This process make sense, however generating basic blocks doesn't seem to be allowed in SPIRV backend yet, as I see this error when creating new MBBs in `llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp`.

```
LLVM ERROR: inserting blocks is not supported yet (in function: test_scalar)
```

Reply:
As discussed, this process should be handles in `llvm/lib/Target/SPIRV/SPIRVISelLowering.cpp`, however, since we don't wanna add new patterns to the intrinsic code gen, I will resolve this discussion.